### PR TITLE
feat: v2 API 👨🏻‍🍳 - V2 contrib evolution

### PR DESCRIPTION
**Pull request by jpmcb**
## Description

This implements a much faster v2 contributor evolution histogram:

---

- Updates the missing v2 for swagger generation
- Updates the `DbContributionStatTimeframe` type to have all the new data points
- Implements a `findContributorsByType` helper function which returns usernames for the given "type" (churn, new, alum, etc.) This was _much_ easier than trying to implement this across a `time_bucket` which I'm not sure is possible. Or at least, it'd be very gross to try to do without a helper query/function
- Implements `findContributionsInTimeFrame` which uses `UNION ALL` across many of the events tables given the usernames

Example `v2/lists/{id}/stats/contributions-evolution-by-type`:

```json5
[
  {
    "bucket": "2024-01-08T00:00:00.000Z",
    "commits": 19,
    "prs_created": 1,
    "prs_reviewed": 17,
    "issues_created": 6,
    "commit_comments": 0,
    "issue_comments": 17,
    "pr_review_comments": 5,
    "comments": 22,
    "total_contributions": 65
  },

  // ... redacted

  {
    "bucket": "2023-12-14T00:00:00.000Z",
    "commits": 3,
    "prs_created": 0,
    "prs_reviewed": 9,
    "issues_created": 1,
    "commit_comments": 0,
    "issue_comments": 4,
    "pr_review_comments": 4,
    "comments": 8,
    "total_contributions": 21
  }
]
```

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents

Related to v2 changes:
- https://github.com/open-sauced/engineering/issues/217
- https://github.com/open-sauced/api/issues/478

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [ ] 🙅 no documentation needed

### CODE REVIEW

File: src/user-lists/entities/contributions-timeframe.entity.ts [hunk 1]
```diff
@@ -75,6 +63,42 @@ export class DbContributionStatTimeframe {
   })
   issues_created: number;
 
+  @ApiModelProperty({
+    description: "",
```
Comment (jpmcb) [inline src/user-lists/entities/contributions-timeframe.entity.ts:67]: TODO: These need descriptions

File: src/user-lists/user-list-events-stats.service.ts [hunk 2]
```diff
@@ -222,6 +295,103 @@ export class UserListEventsStatsService {
...
+    if (allUsers.length === 0) {
+      return new Array<DbContributionStatTimeframe>();
+    }
+
+    const usersWhereConditions = allUsers.map((user) => `LOWER(actor_login) = LOWER('${user}')`).join(" OR ");
+
+    const cteQuery = `
+      SELECT event_type, event_time
+      FROM push_github_events
+      WHERE (${usersWhereConditions})
```
Comment (jpmcb) [inline src/user-lists/user-list-events-stats.service.ts:316]: There's probably a better way to do this with the CTE and using the `setParameters` function

**Review by jpmcb — COMMENTED** [review]
Marking this as draft for now and coming back to it after the weekend

**Comment by jpmcb** [comment]
This is ready to rock and roll into `alpha`: force pushed to add descriptions to the virtual entity and addressed using parameters instead of raw `user` strings stuck together in the `WHERE` for the union tables.

File: src/user-lists/entities/contributions-timeframe.entity.ts [hunk 3]
```diff
@@ -4,31 +4,31 @@ import { Column, Entity } from "typeorm";
 @Entity({ name: "user_list_contributors" })
 export class DbContributionStatTimeframe {
   @ApiModelProperty({
-    description: "The ISO timestamp for the start of the time frame",
-    example: "2023-08-26T23:55:49.204Z",
+    description: "The ISO timestamp for the given time bucket",
```
Comment (nickytonline) [inline src/user-lists/entities/contributions-timeframe.entity.ts:7]: Curious about the wording change. Doesn't it mean the same thing and a time bucket is the implementation of a time frame?

File: src/user-lists/user-list-events-stats.service.ts [hunk 4]
```diff
@@ -39,6 +41,77 @@ export class UserListEventsStatsService {
...
+    const entityQb = this.pullRequestGithubEventsRepository.manager
+      .createQueryBuilder()
+      .addCommonTableExpression(userListQueryBuilder, "CTE")
+      .setParameters(userListQueryBuilder.getParameters())
+      .select("login")
+      .from("CTE", "CTE");
+
+    const entities = await entityQb.getRawMany<{ login: string }>();
+
+    return entities.map((result) => result.login);
```
Comment (nickytonline) [inline src/user-lists/user-list-events-stats.service.ts:112]: IS the map necessary if we're getting back an array of `{ login: string }`? or is `getRawMany` returning more than logins?

File: src/user-lists/entities/contributions-timeframe.entity.ts (hunk 3 above)
Comment (jpmcb) [reply src/user-lists/entities/contributions-timeframe.entity.ts:7]: It is different: the `time_bucket` function in timescale doesn't return the iso timestamp for the _start of the timebucket_, it uses a more complex "origin" system that determines the start of the bucket on sliding windows from the first possible timestamped data.

> By default, a time bucket doesn't start at the earliest timestamp in your data. There is often a more logical time ... time is divided into buckets based on intervals from the origin.

https://docs.timescale.com/use-timescale/latest/time-buckets/about-time-buckets/ 

TLDR: Timescale does intelligent bucketing based on what data it is presented with. These are not necessarily the beginnings of the "data" per-say, which is why you might see drifts from what range you originally provided, but rather, the start of the bucket for what logical data fits into those buckets.

I'm honestly not sure I fully understand it yet, but this is a good thing for us not having to create complex software (like we had before) that looks at data and tries to create buckets based on it.

---

I can go back on this documentation but it made sense to remove the notion of "beginning" since that seemed more complex than just calling it the timestamp of the bucket.

File: src/user-lists/user-list-events-stats.service.ts (hunk 4 above)
Comment (jpmcb) [reply src/user-lists/user-list-events-stats.service.ts:112]: Yeah, I originally tried that, but `getRawMany` with the type cast creates an array of objects whose keys are `login` - this is needed since we are selecting for the `login` column in the query.

I didn't want to have to export that `{ login: string }` type since it's just a one off to get an array of users. So it seemed simpler to just map it out to an array.

**Review by nickytonline — APPROVED** [review]
:shipit:

**Review by brandonroberts — APPROVED** [review]
